src/autocomplete/shared/compute-token-position.ts

```diff
@@ -0,0 +1,117 @@
+import {ParseTree, ParserRuleContext, TerminalNode, Token, TokenStream} from 'antlr4ng';
+
+import {CursorPosition} from './autocomplete-types';
+
+type TokenPosition = {index: number; context: ParseTree; text: string};
+
+export function computeTokenPosition(
+    parseTree: ParseTree | null,
+    tokens: TokenStream,
+    caretPosition: CursorPosition,
+    identifierTokenTypes: number[] = [],
+): TokenPosition | undefined {
+    if (!parseTree) {
+        return;
+    }
+    if (parseTree instanceof TerminalNode) {
+        return computeTokenPositionOfTerminal(
+            parseTree,
+            tokens,
+            caretPosition,
+            identifierTokenTypes,
+        );
+    } else {
+        return computeTokenPositionOfChildNode(
+            parseTree as ParserRuleContext,
+            tokens,
+            caretPosition,
+            identifierTokenTypes,
+        );
+    }
+}
+
+function positionOfToken(
```

Comment:
this is not a proper function name
I suggest renaming this to `computeTokenPositionOfNode` or sth similar